src/traffic_cache_tool/CacheDefs.h

```diff
@@ -275,14 +275,45 @@ class CacheURL
 
 class DFA;
 // this class matches url of the format : scheme://hostname:port/path;params?query
+
 struct url_matcher {
-  // R"(^https?\:\/\/^[a-z A-Z 0-9]\.[a-z A-Z 0-9 \.]+)"
+  url_matcher(ts::FilePath const &path) // file contains a list of regex
+  {
+    ts::BulkFile cfile(path);
+    if (cfile.load() == 0) {
+      ts::TextView fileContent = cfile.content();
+      const char **patterns;
+      std::vector<std::string> str_vec;
+      int count = 0;
+      while (fileContent) {
+        ts::TextView line = fileContent.take_prefix_at('\n');
+        std::string reg_str(line.data(), line.size());
+        str_vec.push_back(reg_str);
+        count++;
+      }
+      patterns = (const char **)ats_malloc(count * sizeof(char *));
+      int i    = 0;
+      for (auto str : str_vec) {
+        patterns[i++] = ats_strdup(str.data());
```

Comment:
Why not `patterns[i++] = str.c_str()`? If you need the `patterns` array to do the regex compilation you can still use `std::string` to hold the long term memory. I'd also be tempted to do something like
```
std::vector<char*> patterns;
patterns.reserve(count);
for (auto& str : str_vec) patterns.push_back(str.c_str);
regex.compile(&(patterns[0]), count);
```

Reply:
will try that